File: schedule_frontend/src/api/timer.ts
import { get, post } from '@/api/client'
import type { TimerOperationData, TimerState } from '@/types/timer'

export function fetchCurrentTimer(): Promise<TimerState> {
  return get<TimerState>('/api/timer/current')
}

export function startTimer(taskId: number): Promise<TimerOperationData> {
  return post<TimerOperationData>('/api/timer/start', { body: { task_id: taskId } })
}

export function pauseTimer(): Promise<TimerOperationData> {
  return post<TimerOperationData>('/api/timer/pause')
}

export function resumeTimer(): Promise<TimerOperationData> {
  return post<TimerOperationData>('/api/timer/resume')
}

export function stopTimer(note?: string): Promise<TimerOperationData> {
  return post<TimerOperationData>('/api/timer/stop', { body: { note } })
}

export function discardTimer(): Promise<TimerOperationData> {
  return post<TimerOperationData>('/api/timer/discard')
}

export function switchTimer(newTaskId: number, saveCurrentSession = true, note?: string): Promise<TimerOperationData> {
  return post<TimerOperationData>('/api/timer/switch', {
    body: {
      new_task_id: newTaskId,
      save_current_session: saveCurrentSession,
      note,
    },
  })
}

File: schedule_backend/app/api/timer.py
from __future__ import annotations

from fastapi import APIRouter

from app.api.deps import DBSessionDep
from app.core.response import ApiResponse, success_response
from app.schemas.timer import TimerCurrentData, TimerOperationData, TimerStartRequest, TimerStopRequest, TimerSwitchRequest
from app.services.timer_service import TimerService

router = APIRouter(prefix="/timer", tags=["timer"])


@router.get("/current", response_model=ApiResponse[TimerCurrentData])
def get_current_timer(db: DBSessionDep) -> ApiResponse[TimerCurrentData]:
    data = TimerService(db).get_current_timer()
    return success_response(data)


@router.post("/start", response_model=ApiResponse[TimerOperationData])
def start_timer(payload: TimerStartRequest, db: DBSessionDep) -> ApiResponse[TimerOperationData]:
    data = TimerService(db).start_timer(payload)
    return success_response(data)


@router.post("/pause", response_model=ApiResponse[TimerOperationData])
def pause_timer(db: DBSessionDep) -> ApiResponse[TimerOperationData]:
    data = TimerService(db).pause_timer()
    return success_response(data)


@router.post("/resume", response_model=ApiResponse[TimerOperationData])
def resume_timer(db: DBSessionDep) -> ApiResponse[TimerOperationData]:
    data = TimerService(db).resume_timer()
    return success_response(data)


@router.post("/stop", response_model=ApiResponse[TimerOperationData])
def stop_timer(payload: TimerStopRequest, db: DBSessionDep) -> ApiResponse[TimerOperationData]:
    data = TimerService(db).stop_timer(payload)
    return success_response(data)


@router.post("/discard", response_model=ApiResponse[TimerOperationData])
def discard_timer(db: DBSessionDep) -> ApiResponse[TimerOperationData]:
    data = TimerService(db).discard_timer()
    return success_response(data)


@router.post("/switch", response_model=ApiResponse[TimerOperationData])
def switch_timer(payload: TimerSwitchRequest, db: DBSessionDep) -> ApiResponse[TimerOperationData]:
    data = TimerService(db).switch_timer(payload)
    return success_response(data)
